src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,211 @@
+                        ips->push_back(answer.ipv4);
+                    } else if (answer.hostname != "") {
+                        logger->info("(2) %s ipv4: %s",
+                            hostname.c_str(), answer.ipv4.c_str());
+                        ips->push_back(answer.ipv4);
+                    }
+                }
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                logger->info("dns_many() found %d ips", (int)ips->size());
+                cb(NoError(), *ips);
+            }
+        };
+    };
+
+    for (auto const& hostname : hostnames) {
+        Settings options; // XXX ought this to forward from above?
+        // XXX don't hardcode resolver
+        templates::dns_query(entry, "A", "IN", hostname, "8.8.8.8",
+                             dns_cb(hostname), options, reactor, logger);
+    }
+    return;
+}
+
+static void http_many(const std::vector<std::string> urls,
+                      Var<Entry> entry,
+                      Callback<Error> cb,
+                      Var<Reactor> reactor,
+                      Var<Logger> logger) {
+    int urls_count = urls.size();
+    Var<int> urls_tested(new int(0));
+    if (urls_count == *urls_tested) {
+        cb(NoError());
+        return;
+    }
+
+    auto http_cb = [=](std::string url) {
+        return [=](Error err, Var<http::Response> response) {
+             Entry result = {
+                 {"url", url},
+                 {"status",
+                     {{"success", nullptr},
+                     {"failure", nullptr}}},
+             };
+             if (!!err) {
+                 logger->info("whatsapp: failure HTTP connecting to %s",
+                     url.c_str());
+                 result["status"]["success"] = false;
+                 result["status"]["failure"] = err.as_ooni_error();
+             } else {
+                 logger->info("whatsapp: success HTTP connecting to %s",
+                     url.c_str());
+                 result["status"]["success"] = true;
+             }
+            (*entry)["http_connect"].push_back(result);
+             *urls_tested += 1;
+             if (urls_count == *urls_tested) {
+                 cb(NoError());
+             }
+         };
+    };
+
+    for (auto url : urls) {
+        Settings http_options;
+        http_options["http/url"] = url;
+        http::Headers headers = constants::COMMON_CLIENT_HEADERS;
+        std::string body;
+        templates::http_request(entry, http_options, headers, body,
+                                http_cb(url), reactor, logger);
+    }
+
+}
+
+void whatsapp(std::string input, Settings options,
+              Callback<Var<report::Entry>> callback,
+              Var<Reactor> reactor, Var<Logger> logger) {
+    std::vector<std::string>
+        WHATSAPP_WEB_URLS = { "https://web.whatsapp.com/",
+                              "http://web.whatsapp.com/" };
+    // XXX generate these with a range or something
```

Comment:
Ideally, I guess that this can be some sort of input file. (Now that you have implemented all the IM tests, I see that they have lots in common and they mostly differ for their input. I also see that there are a number of places where we can benefit from adding more functional abstractions to the code, and perhaps we should do that so to simplify coding.)